plat/drivers: Provide a single API for UART drivers

zhxj9823 opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): `x86_64` and `ARM64`
 - Platform(s): `kvm`
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This PR migrates the console API into a new `libukconsole`. Each console provides an ops structure that implements the console API. The APIs include `coutk`, `coutd`, `cink`, and `init`.

A unified `uk_console_init` function takes `bootinfo` as a parameter to initialize the available console in both x86 and ARM platforms. The implementation of `uk_console_init` is platform-specific.

The header for ANSI characters is moved into a separate one in `include/uk`.

GitHub-Depends-On: #989

mogasergiu committed:
PR #909: Add EFI stub support

mogasergiu committed:
PR #772: Add PIE support for Xen and KVM

mogasergiu committed:
PR #848: Integrate ukplat_memregion into Xen x86, AArch64 KVM and linuxu

mogasergiu committed:
PR #908: Add ukvmem Aarch64 page fault handler

mogasergiu committed:
PR #911: Refactor the ACPI code and change its location

mogasergiu committed:
PR #910: Add the ability to build EFI ISO and DISK images

mogasergiu committed:
PR #912:  Add support for setting up SMP on ARM64 through ACPI

mogasergiu committed:
PR #907: arch/arm/arm64: Fix stxr's used in spinlocks register size

michpappas committed:
plat/kvm/arm: Add pagetables for Firecracker

The memory map of Firecracker on Arm64 is:

0x0000000000000000 - 0x0000000007ffffff     Devices:     0 -    2GiB
0x0000000080000000 - 0x000000ffffffffff     DRAM:     2GiB - 1024GiB

https://github.com/firecracker-microvm/firecracker
441fc98a src/vmm/src/arch/aarch64/layout.rs

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
PR unikraft#988: Implement lxboot on arm64

michpappas committed:
plat/kvm: Enable Firecracker on arm64

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

razvand commented:
Hi, @zhxj9823, this is marked as draft work. What needs to happen to make this a full PR?

razvand commented on lib/ukconsole/console.c:
```diff
@@ -27,12 +28,67 @@
+
+	if ((offset = fdt_node_offset_by_compatible(dtb, -1, "ns16550")) >= 0
+	    && (offset = fdt_node_offset_by_compatible(dtb, -1, "ns16550a"))
+		   >= 0) {
```
See comment from Codacy. Where is offset used?

razvand reviewed (changes requested):
For the last two commits of this PR, please add a motivation for their existence. Such as why move the x86 console to a COM driver under `plat/drivers/uart/`.

Also, see some minor typos: `migrate` -> `migrates`, `an addition` -> `an additional`.

Also it would be ideal to add tests as part of the new `ukconsole` library, before upstreaming it, via the [`uktest` framework](https://unikraft.org/docs/develop/writing-tests/). @michpappas , what do you think? I'm OK with upstreaming it as part of the 0.14 release even if the tests aren't there.

zhxj9823 commented:
>> Also it would be ideal to add tests as part of the new `ukconsole` library, before upstreaming it, via the [`uktest` framework](https://unikraft.org/docs/develop/writing-tests/). @michpappas , what do you think? I'm OK with upstreaming it as part of the 0.14 release even if the tests aren't there.

> The input and output of this library are from and to the serial console. Can this be fitted into the `uktest` framework?

Does not fit. In general devices are a bit more complex to write tests for, we would need to introduce some abstractions so it would take time. I propose that we skip it for now. @razvand.

michpappas commented on lib/ukconsole/Config.uk:
```diff
@@ -0,0 +1,68 @@
+
+if LIBUKCONSOLE
+
+config KVM_KERNEL_SERIAL_CONSOLE
```
KVM configuration should be in `plat/kvm/Config.uk`. The location of driver-specific configuraiton is a bit tricky right now because we don't have not moved drivers out of `plat` yet, so there is a gap on where these should go. Maybe you can move it back to `plat/kvm` for now, and I'll try to come back with an update.

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -27,12 +28,67 @@
+	return uk_console_ops->cink(buf, maxlen);
+}
+
+#if CONFIG_ARCH_X86_64
```
If you need per-architecture variants you need to place architecture-specific variants under `lib/ukconsole/arch/<arch>/<name>.c` and update your `Makefile.uk` accordingly.

zhxj9823 replied:
In this case, would it make sense to divide the `.h` and `.c` files? For the `console.h` file, the ANSI characters and console APIs are used in different cases. For the `console.c` file, it can be divided into the arch-specific `console_init` function and the other three arch-independent functions.

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -1,6 +1,7 @@
 /* SPDX-License-Identifier: BSD-3-Clause */
 /*
- * Copyright (c) 2021 OpenSynergy GmbH
```
Here it looks like derived work from an existing file, which is not the case. Please introduce a new file. The copyright for new files should be:
```
/* SPDX-License-Identifier: BSD-3-Clause */
/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
 * Licensed under the BSD-3-Clause License (the "License").
 * You may not use this file except in compliance with the License.
 */
```

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -27,12 +28,67 @@
+extern const struct uk_console_ops uk_console_pl011_ops;
+extern const struct uk_console_ops uk_console_ns16550_ops;
+
+void uk_console_init(const void *dtb)
```
To avoid having different function prototypes per architecture, and to support in the future other discovery methods like ACPI, you can instead pass all instances a parameter to the bootinfo. In the case of the dtb you should get the address from there.

As discussed, that depends on #912, so please rebase and mark the dependency in the cover letter of your PR using `GitHub-Depends-On`. For reference you can see https://github.com/unikraft/unikraft/pull/988 and https://github.com/unikraft/unikraft/pull/989.

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -100,6 +100,17 @@ extern "C" {
+	int (*coutk)(const char *buf, unsigned int len);
+	int (*coutd)(const char *buf, unsigned int len);
+	int (*cink)(char *buf, unsigned int maxlen);
+#if CONFIG_ARCH_X86_64
```
To avoid having different function prototypes per architecture, and to support in the future other discovery methods like ACPI, you can instead pass all instances a parameter to the bootinfo. In the case of the dtb you should get the address from there.

As discussed, that depends on #912, so please rebase and mark the dependency in the cover letter of your PR using `GitHub-Depends-On`. For reference you can see https://github.com/unikraft/unikraft/pull/988 and https://github.com/unikraft/unikraft/pull/989.

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -27,12 +28,67 @@
+
+static struct uk_console_ops *uk_console_ops;
+
+int ukplat_coutk(const char *buf, unsigned int len)
```
These should no longer be `ukplat_`  but `uk_console`

michpappas commented on plat/kvm/Makefile.uk:
```diff
@@ -82,14 +82,15 @@ ifeq ($(findstring y,$(CONFIG_KVM_KERNEL_VGA_CONSOLE) $(CONFIG_KVM_DEBUG_VGA_CON
 ##
 ifeq ($(findstring y,$(CONFIG_KVM_KERNEL_SERIAL_CONSOLE) $(CONFIG_KVM_DEBUG_SERIAL_CONSOLE)),y)
 LIBKVMPLAT_SRCS-$(CONFIG_ARCH_ARM_64) += $(UK_PLAT_DRIVERS_BASE)/uart/pl011.c
+LIBKVMPLAT_SRCS-$(CONFIG_ARCH_ARM_64) += $(UK_PLAT_DRIVERS_BASE)/uart/ns16550.c
```
Looks like this change does not belong to this commit.

michpappas commented on plat/drivers/include/uart/com.h:
```diff
@@ -19,11 +19,11 @@
+#ifndef __COM_H__
+#define __COM_H__
 
 void _libkvmplat_init_serial_console(void);
```
COMx devices are not KVM specific: https://en.wikipedia.org/wiki/COM_(hardware_interface), so the interface provided by this driver should come with appropriate names. Also shouldn't this driver also provide a set of ops like the rest?

zhxj9823 replied:
This relates to the issue that `vga` console and `com` console are used concurrently in `unikraft/plat/kvm/x86/console.c`. In the current scheme, there can be an `ops` structure for this file. However, having a separate `ops` structure for `com` driver breaks the backward compatibility. Another choice is to enable multiple consoles instead of only using one.

michpappas commented on plat/drivers/uart/ns16550.c:
```diff
@@ -173,14 +168,14 @@ static int ns16550_getc(void)
 }
 
-int ukplat_coutk(const char *buf, unsigned int len)
+int ns16550_coutk(const char *buf, unsigned int len)
```
These should be `static` now. In the rest of the drivers too.

michpappas reviewed (commented):
Hi @zhxj9823 thanks for this work 🚀 I did a first pass, it's likely that more iterations will follow. 

Besides the inline comments, according to Unikraft's guidelines changes to each library must come on a separate commit. This applies to any updates to `nolibc`, `ukdebug`, and `ukboot`. Moreover, please introduce the `ukconsole` library on a separate commit too. Similarly for changes on each driver.

zhxj9823 committed:
lib/ukconsole: Add library for serial console

This commit migrates the console API into a new libukconsole.
Each console provides an ops structure that implements the console API.
The APIs include coutk, coutd, cink, and init.

A unified uk_console_init function takes bootinfo as a parameter to
initialize the available console in both x86 and ARM platforms. The
implementation of uk_console_init is platform-specific.

The header for ANSI characters is moved into a separate one in
include/uk.

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
plat/drivers: Migrate pl011 into ukconsole APIs

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
plat/drivers: Migrate ns16550 into ukconsole APIs

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
plat/drivers: Migrate x86 console into com driver

The x86 console from plat/kvm/x86/serial_console.c is moved
into a com driver under plat/drivers/uart.

The APIs for console are migrated into ukconsole APIs.

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
lib/uktest: Change to new header file

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
lib/devfs: Use ukconsole APIs

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
lib/nolibc: Use ukconsole APIs

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
lib/syscall_shim: Migrate to ukconsole APIs

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
lib/ukboot: Change to new header file

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
lib/ukdebug: Migrate to ukconsole APIs

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
lib/ukrust: Change to new header file

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

zhxj9823 committed:
lib/vfscore: Use ukconsole APIs

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

michpappas commented:
I had a brief look on how the reg-shift and width for the `earlycon` are specified on linux. There these are not compiled in, but rather derived from the command-line options:
* https://www.kernel.org/doc/html/v4.14/admin-guide/kernel-parameters.html
* https://elixir.bootlin.com/linux/latest/source/drivers/tty/serial/earlycon.c#L93

The latter derives the regshift from the mmio width, ie 0 by default, 1 for MMIO16, and 2 for MMIO32 access. Right now we use 16-bit access and a regshift of 2 which doesn't make sense. @kubanrob should that be 32-bit access instead? Also would the command line be a valid option for earlycon in your case, assuming we implemented that? I imagine that since it's used on linux, it should be no problem for Unikraft.

zhxj9823 committed:
plat/drivers: Add config for selecting consoles

This commit adds configurations for selecting which console to use. When
a specific console is selected, it can be configured as desired.

The early debug console is also handled. A uk_console_null_ops is added
to handle the case where no early debug console is available.

Signed-off-by: Xingjian Zhang <zhxj9823@qq.com>

michpappas commented:
Hi @zhxj9823. I opened a PR with some changes to introduce a drivers subsystem #1023, and migrated the `pl011` and `ns16550` drivers there (#1025 and #1026). You can update this PR to move the COM and VGA console drivers to `drivers/uktty` using `pl011` as a template. Moveover, please rename this library from `ukconsole` to `uktty`.

You can also consider refactoring the current PR to register the `pl011` and `ns16550` drivers as platform drivers using the `PF_REGISTER_DRIVER` macro of the the platform bus (see `plat/common/include/platform_bus.h`). That will allow you to remove the dts probing from this library and replace it with a `uktty_init()` that the drivers will call on `add_device` to pass their ops to the `uktty` library. Notice that this deprecates the architecture-specific parts from this library.

michpappas commented:
Obsoleted by #1464
